### PR TITLE
Rover: allow vehicle to be armed in manual with no GPS

### PR DESCRIPTION
**Pull request by peterbarker**

### CODE REVIEW

**Comment by rmackay9** [comment]
I definitely like the new checks to ensure we've got GPS before going into a mode that requires GPS.  It seems a bit like a lot is duplicated from the AP_Arming checks?  I guess no way to re-use the AP_Arming calls?

.. or maybe we can do something like in Copter where we have a position_ok() function that basically just asks the EKF if it's happy.

File: APMrover2/mode.cpp [hunk 1]
```diff
@@ -19,8 +21,56 @@ void Mode::exit()
...
+                        (double)distance_m);
+        return false;
+    }
+    if (!gps.blend_health_check()) {
+        gcs().send_text(MAV_SEVERITY_CRITICAL, "GPS blending unhealthy");
+        return false;
+    }
+
+    // check AHRS and GPS are within 10m of each other
+    Location gps_loc = gps.location();
```
Comment (khancyr) [inline APMrover2/mode.cpp]: const

File: APMrover2/mode.cpp [hunk 2]
```diff
@@ -19,8 +21,56 @@ void Mode::exit()
...
+    Location ahrs_loc;
+    if (ahrs.get_position(ahrs_loc)) {
+        float distance = location_3d_diff_NED(gps_loc, ahrs_loc).length();
```
Comment (khancyr) [inline APMrover2/mode.cpp]: const

**Comment by khancyr** [comment]
I am not sure about this check as we can use visual odometry or wheel odom to move without gps

**Comment by peterbarker** [comment]
On Wed, 29 Nov 2017, Randy Mackay wrote:

> I definitely like the new checks to ensure we've got GPS before going into a
> mode that requires GPS. It seems a bit like a lot is duplicated from the
> AP_Arming checks? I guess no way to re-use the AP_Arming calls?

Yes, I even commented on the duplication.

I actually tried directly calling the arming checks!  The problem lies in 
the new check which allows arming in non-gps mode; it is based on the 
current mode, not the mode you're going into - so you end up skipping the 
checks if you try to move from manual -> steering, which is not optimal.

> .. or maybe we can do something like in Copter where we have a position_ok()
> function that basically just asks the EKF if it's happy.

I'm happy to simplify this to a position_ok; that's what I started to do.

I'll just paste in some of my musings from Skype here for posterity:

---


More I think about this the more I think it's all wrong.   manual, 
steering (and probably several other modes) shouldn't require GPS.  But it 
is almost certainly the case that those other modes do require some form 
of speed detection - GPS, Optical Flow, wheel encoders, whatever.   We 
should probably break the concepts out - have_position_global(), 
have_position_local() and maybe have_speed_estimate().

Steering doesn't require global position but does require a speed 
estimate.

**Comment by peterbarker** [comment]
On Wed, 29 Nov 2017, Pierre Kancir wrote:

> I am not sure about this check as we can use visual odometry or wheel odom
> to move without gps

Yes!  Please see (amd comment on!) my musings elsewhere on this PR.

In this case we're just loosening up the current restriction on manual 
mode, however - so while yes, I agree, things could be much, much better, 
we don't need to get there on this PR :-)

File: APMrover2/mode.cpp (hunk 1 above)
Comment (rmackay9) [reply APMrover2/mode.cpp]: fixed before merging.

File: APMrover2/mode.cpp (hunk 2 above)
Comment (rmackay9) [reply APMrover2/mode.cpp]: oops, forgot this one before i merged!

**Comment by rmackay9** [comment]
OK, merged!

**Pull request closed without merging**